crates/ide_db/src/path_transform.rs

```diff
@@ -145,10 +151,60 @@ impl<'a> Ctx<'a> {
         match resolution {
             hir::PathResolution::TypeParam(tp) => {
                 if let Some(subst) = self.substs.get(&tp) {
-                    ted::replace(path.syntax(), subst.clone_subtree().clone_for_update().syntax())
+                    let parent = path.syntax().parent()?;
+                    if let Some(parent) = ast::Path::cast(parent.clone()) {
+                        // Path inside path means that there is an associated
+                        // type on the type parameter. It is necessary to fully
+                        // qualify the type with `as Trait`. Even though it
+                        // might be unnecessary if `subst` is generic type,
+                        // always fully qualifying the path is safer because of
+                        // potential clash of associated types from multiple
+                        // traits
+
+                        let trait_ref = find_trait_for_assoc_type(
+                            self.source_scope,
+                            tp,
+                            parent.segment()?.name_ref()?,
+                        )
+                        .and_then(|trait_ref| {
+                            let found_path = self.target_module.find_use_path(
+                                self.source_scope.db.upcast(),
+                                hir::ModuleDef::Trait(trait_ref),
+                            )?;
+                            match ast::make::ty_path(mod_path_to_ast(&found_path)) {
+                                ast::Type::PathType(path_ty) => Some(path_ty),
+                                _ => None,
+                            }
+                        });
+
+                        let segment = ast::make::path_segment_ty(subst.clone(), trait_ref);
+                        let qualified =
+                            ast::make::path_from_segments(std::iter::once(segment), false);
+                        ted::replace(path.syntax(), qualified.clone_for_update().syntax());
+                    } else if let Some(path_ty) = ast::PathType::cast(parent) {
+                        ted::replace(
+                            path_ty.syntax(),
+                            subst.clone_subtree().clone_for_update().syntax(),
+                        );
+                    } else {
+                        ted::replace(
```

Comment:
Even without this `else` branch, all tests pass. I am wondering in which scenarios is this code executed and if that's what is desired.

Reply:
The only case where I can see this happening is in Patterns using the path I believe, so `PathPat`, or the path in `TupleStructPat` etc, might be good to check that.

Reply:
I couldn't come up with a code snippet that would exercise this branch. It would need to:

* use a generic type parameter, because this part of code deals with `hir::PathResolution::TypeParam(tp)` substitution
* must not use associated type/constant, because this case is handled by `if let Some(parent) = ast::Path::cast(parent.clone())` above
* must be in a pattern or path expression

Therefore, as far as I can tell, the path here will always be a type (having `PathType` as parent) or will have an associated type/constant (having `Path` as parent). And so this `else` branch seems unreachable to me and can be removed.

Nevertheless, thanks to playing with path patterns and expressions, I discovered that handling associated *constants* was missing in my fix. It works now (d9b3242bcd5f4208fad707a30907633ccf081056).

Reply:
Can't hurt leaving the branch either so I think keeping it is fine, either way works.

PR lgtm then, thanks!
bors d+